### ISSUE TITLE
[Question] Why not use edge.js?

### ISSUE THREAD

**Issue by donatasbacius** [reactjs/React.NET#134]
Hi @Daniel15,

What do you think of this approach? Why not delegate the rendering to node.js? That way you get commonjs support and can use pretty much any node modules with minimal effort.

Thanks

**Comment by Daniel15** [comment]
Does that require Node.js to be installed? That's a higher barrier to entry than what I'd like 'out of the box', but it'd be good if it was optional. When I started working on ReactJS.NET, edge.js didn't support Linux (via Mono) which was one of my goals for 1.0. It looks like it does support Mono now, so it might be worth revisiting.

For what it's worth, CommonJS support is already possible if you use Webpack or Browserify to create a bundle that ReactJS.NET uses.

**Comment by andersekdahl** [comment]
We've started to use Edge.js for server rendering React, and so far so good. It doesn't require you to install Node.js, it bundles Node.js with the NuGet package. All APIs to communicate with Edge are aync from the C# side which is nice, but it makes it impossible to use Edge from an HtmlHelper (which is fixed in MVC6 with tag helpers).

Another thing to be aware of is that there can only be one Node/Edge instance/thread per IIS process, so you need to change the number of IIS workers for it to run in parallell.

**Comment by PeteDuncanson** [comment]
Personally I don't see whats wrong with simply using Webpack to bundle up your scripts. Means you get server-side and client-side all done in one step. I don't know what you would be gaining by doing it Edge?

**Comment by andersekdahl** [comment]
We're actually using Webpack to build a bundle for server rendering that we `require` with Edge.js.

We're still in early stages of server rendering React, and the interesting options are this project, Edge.js and [NodeServices](https://github.com/aspnet/NodeServices) where NodeServices is just an early proof of concept, but it's interesting to see Microsoft exploring this.

When we have a working site with SSR and Edge.js, I intend to benchmark it and then try out React.NET to compare them. I'd also be very interested in testing SuperchargedReact.Net, but I'm not sure how far you have come with that project?

**Comment by PeteDuncanson** [comment]
@andersekdahl I'm quite open about SuperCharged in that its come far enough to meet our needs so attention has moved on to using it rather than developing it plus it means currently its not all that user friendly for newbies (no documentation!) even though there is not much to it really. 

However I'm very keen to get more people having a play with it. Sadly paying the mortgage and doing up the house and playing with my kid mean evening coding time is zero these day. However I do know it need some more documentation to get folks up and running with it quicker and easier. Its not doing any black magic, it just does away with some of the constraints Daniel had in React.net (or added some moreconstraints I'm not sure which way around that should be), for instance we only work with V8 and I've not even given a moments thought about Mono or Asp.net V5 etc :)

What I can say is its doing a sterling job on a few of our client sites and I'd happily use it on more in the future.

**Comment by andersekdahl** [comment]
@PeteDuncanson That's very interesting! Documentation would of course be nice, but I prefer digging around in code anyway.
What kind of load do you have on the sites you're using it on?

Are these repos up to date and a good starting point to test it out?
- https://github.com/Offroadcode/Umbraco-React-Starter-kit
- https://github.com/Offroadcode/SuperchargedReact.Net

**Comment by PeteDuncanson** [comment]
@andersea yep thats the latest. You can either run it with ReactRouter or to render a single component via a HtmlHelper.

General gist is you need to include reactrunner.js in your project and that does the magic of working out if you are rendering using ReactRouter or not. Globally we have a setting object that SuperCharged pokes into the JS engine with the settings from the server if you need to pass anything extra in. Off the top of my head those are the two main "tricky bits" that folks see as "magic" but they work pretty damn good they just need documenting!

**Comment by andersea** [comment]
I was summoned. 8-)

**Comment by PeteDuncanson** [comment]
Haha! Oops, sorry @andersea - I blame autocorrect :)

**Comment by andersekdahl** [comment]
I've statred to play with it, having some issues with the ClearScript dlls but I'll figure it out.
What kind of load does the sites using SuperChargedReact.Net have? Number of concurrent users?

**Comment by PeteDuncanson** [comment]
The ClearScript DLL's are a pain, you need to have them in two place:

/ClearScript.V8
 - ClearScriptV8-32.dll
 - ClearScriptV8-64.dll
 - v8-ia32.dll
 - v8-x64.dll
/bin/
 - ClearScript.dll
 - v8-x64.dll (or v8-x32.dll depending on your environment)

I'll admit I've been a bit cargo cult with these, I don't really understand how they work or why they have to be where they have to be and there might be a better way to do it, etc. etc. but it works like this which lets me concentrate on other more fun things. If you find a better way let me know! :)

**Comment by andersekdahl** [comment]
Thanks for the pointers!
You never answered my question about the load on your sites though. :)

**Comment by PeteDuncanson** [comment]
@andersekdahl Ooops sorry, task focused! We peak out just shy of 100 users per second currently and on Black Friday we had nearly 600 active sessions at once (not 600 hits at once though obviously). Its not totally crazy but for our needs it does just fine. CPU's on the servers barely flicker too :)

This is not however an all singing and dancing app with millions of users but like I said for our needs its doing great.

**Comment by jslatts** [comment]
@PeteDuncanson It's good to hear you are using it in production. Dropping SuperChargedReact.NET into our app for testing is on my todo list as well. Do you guys hang out on IRC? Would love to chat with you a bit more about your project.

**Comment by PeteDuncanson** [comment]
@jslatts I've never been one for IRC, I was on the React slack channel for a bit but it soon turns into a long chat about what everyone is having for breakfast every morning and of course the constant updates about coffee (it should be updates about tea of course but each to their own).

**Comment by Daniel15** [comment]
I do still want to take some ideas from SuperchargedReact.NET for ReactJS.NET, I just haven't had time to do so. :+1: 

> and NodeServices where NodeServices is just an early proof of concept, but it's interesting to see Microsoft exploring this.

That's indeed very interesting, I hadn't heard about this until today. I didn't know that Microsoft were working on something similar :open_mouth: 

>  I've never been one for IRC, I was on the React slack channel for a bit but it soon turns into a long chat about what everyone is having for breakfast every morning

lol

The Slack chat moved to a different system called Discord. I'll see if they can create a `#dotnet` channel to chat about all things React + ASP.NET :smile:

**Comment by jslatts** [comment]
+1 on the discord channel. It would give me a reason to hang out there.

**Comment by andersekdahl** [comment]
I've never used discord but I'd definitely join if there was a channel to talk about React + ASP.NET!

**Comment by Daniel15** [comment]
There's a Gitter channel, it's a bit... dead though. I think it's just me
in there now :P

https://gitter.im/reactjs/React.NET

--
Regards,
Daniel Lo Nigro
http://dan.cx/ | Twitter <http://twitter.com/Daniel15> | Facebook
<http://www.facebook.com/daaniel>

**Comment by pauldotknopf** [comment]
I looked into using edge.js with my react asp.net starter template, [react-aspnet-boilerplate](https://github.com/pauldotknopf/react-aspnet-boilerplate). It only allowed a single instance, and added a lot of technical overhead for functionality that really isn't needed. We should just depend on using scripts compiled by webpack in a traditional javscript environment (V8, ClearScript, etc). That is what I winded up doing.

I wish someone would build some C layer (for P-Invoke) over [atom's node](https://github.com/atom/node) implementation which allows embedding the node.js runtime.

Also, if anybody is looking for another react .net template, checkout [react-aspnet-boilerplate](https://github.com/pauldotknopf/react-aspnet-boilerplate). It implements the default aspnet project functionality (authentication/account/etc), universal rendering, redux, react-router, etc.
